File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D09.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y23D9(input: String) : Day {

    private val patterns = input.lines().map { line -> line.split(' ').map(String::toInt) }

    private fun findNext(pattern: List<Int>) = generateSequence(pattern) { it.zipWithNext { a, b -> b - a } }
        .takeWhile { next -> next.any { it != 0 } }
        .sumOf { it.last() }
    
    override fun part1() = patterns.sumOf { pattern -> findNext(pattern) }

    override fun part2() = patterns.sumOf { pattern -> findNext(pattern.reversed()) }
}

fun main() = Day.runDay(Y23D9::class)

//    Class creation: 17ms
//    Part 1: 1974913025 (6ms)
//    Part 2: 884 (3ms)
//    Total time: 27ms

@Suppress("unused")
private val sampleInput = listOf(
    """0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45""",
)

File: rust/advent/src/bin/2023/y23d09.rs
use std::iter::successors;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = i64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 9).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn patterns(input: Input) -> impl Iterator<Item = Vec<i64>> + use<'_> {
    input
        .lines()
        .map(|line| line.split(' ').map(|n| n.parse().unwrap()).collect())

}

fn find_next(pattern: Vec<i64>) -> i64 {
    successors(Some(pattern), |it| {
        Some(it.iter().tuple_windows().map(|(a, b)| b - a).collect())
    })
    .take_while(|next| next.iter().any(|it| it != &0) )
    .map(|it| *it.last().unwrap())
    .sum()
}

fn part1(input: Input) -> Output {
    patterns(input)
        .map(find_next)
        .sum()
}

fn part2(input: Input) -> Output {
    patterns(input)
        .map(|mut pattern| {
            pattern.reverse();
            find_next(pattern)
        })
        .sum()
}

#[test]
fn default() {
    let input = get_input(23, 9).unwrap();
    assert_eq!(1974913025, part1(&input));
    assert_eq!(884, part2(&input));
}

// Input parsed (20μs)
// 1. 1974913025 (152μs)
// 2. 884 (139μs)
// Total: 314μs